v2/stream/new/client.go

```diff
@@ -0,0 +1,336 @@
+}
+
+var ErrConnectCalledMultipleTimes = errors.New("Connect called multiple times")
+
+// Connect establishes a connection for c using its configuration and reestablishes it when errors occur.
+// It blocks until the connection has been established for the first time (or it failed to do so).
+// Should only be called once!
+//
+// It returns after the connection was established for the first time or if it couldn't be established
+// after retrying.
+func (c *client) Connect(ctx context.Context) error {
+	err := ErrConnectCalledMultipleTimes
+	c.connectOnce.Do(func() {
+		err = c.connect(ctx)
+		if err != nil {
+			c.connErrChan <- err
+		}
+		c.connectCalled = true
+	})
+	return err
+}
+
+func (c *client) connect(ctx context.Context) error {
+	u, err := constructURL(c.host, c.feed)
+	if err != nil {
+		return err
+	}
+
+	successCh := make(chan struct{})
+	go c.maintainConnection(ctx, u, successCh)
+
+	select {
+	case <-successCh:
+		return nil
+	case err := <-c.connErrChan:
+		return err
+	}
+}
+
+func (c *client) Error() <-chan error {
+	return c.connErrChan
+}
+
+// maintainConnection initializes a connection to u, starts the necessary goroutines
+// and recreates them if there was an error as long as reconnectLimit consecutive
+// connection initialization errors don't occur. It closes successCh upon the first
+// successful connection intialization
+func (c *client) maintainConnection(ctx context.Context, u url.URL, successCh chan<- struct{}) {
+	var connError error
+	failedAttemptsInARow := 0
+	connectedAtLeastOnce := false
+
+	defer func() {
+		// If there is a pending sub change we should terminate that
+		c.pendingSubChangeMutex.Lock()
+		defer c.pendingSubChangeMutex.Unlock()
+		if c.pendingSubChange != nil {
+			c.pendingSubChange.result <- ErrSubChangeInterrupted
+		}
+		c.pendingSubChange = nil
+		c.hasTerminated = true
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			if !connectedAtLeastOnce {
+				c.logger.Warnf("datav2stream: cancelled before connection could be established, last error: %v", connError)
+				err := fmt.Errorf("cancelled before connection could be established, last error: %w", connError)
+				c.connErrChan <- err
+			}
+			return
+		default:
+			if c.reconnectLimit != 0 && failedAttemptsInARow >= c.reconnectLimit {
+				c.logger.Errorf("datav2stream: max reconnect limit has been reached, last error: %v", connError)
+				c.connErrChan <- fmt.Errorf("max reconnect limit has been reached, last error: %w", connError)
+				return
+			}
+			time.Sleep(time.Duration(failedAttemptsInARow) * c.reconnectDelay)
+			failedAttemptsInARow++
+			c.logger.Infof("datav2stream: connecting to %s, attempt %d/%d ...", u, failedAttemptsInARow, c.reconnectLimit)
+			conn, err := connCreator(ctx, u)
+			if err != nil {
+				connError = err
+				c.logger.Warnf("datav2stream: failed to connect, error: %v", err)
+				continue
+			}
+			c.conn = conn
+
+			c.logger.Infof("datav2stream: established connection")
```

Comment:
minor: `c.logger.Info`?

Reply:
Now, I see that we have a separate `Logger` interface only with `Infof`, `Warnf`, and `Errorf` methods.

Reply:
Yes, I wanted to keep the interface as small as possible.